bioimageio/core/build_spec/build_spec.py

```diff
@@ -0,0 +1,440 @@
+    )
+    outputs = _get_output_tensor(
+        test_out,
+        output_name,
+        output_reference,
+        output_scale,
+        output_offset,
+        output_axes,
+        output_data_range,
+        postprocessing,
+        halo,
+    )
+
+    (weights, language, framework, source_hash) = _get_weights(weight_uri, weight_type, source, root, **weight_kwargs)
+
+    #
+    # generate general fields
+    #
+    format_version = get_args(model_spec.raw_nodes.FormatVersion)[-1]
+    timestamp = datetime.datetime.now()
+
+    if source is not None:
+        source = spec.shared.fields.ImportableSource().deserialize(source)
+
+    # optional kwargs, don't pass them if none
+    optional_kwargs = {
+        "git_repo": git_repo,
+        "attachments": attachments,
+        "packaged_by": packaged_by,
+        "parent": parent,
+        "run_mode": run_mode,
+        "config": config,
+        "sample_inputs": sample_inputs,
+        "sample_outputs": sample_outputs,
+        "framework": framework,
+        "language": language,
+        "source": source,
+        "sha256": source_hash,
+        "kwargs": model_kwargs,
+        "dependencies": dependencies,
+        "links": links,
+    }
+    kwargs = {k: v for k, v in optional_kwargs.items() if v is not None}
+
+    # build raw_nodes objects
+    authors = _build_authors(authors)
+    cite = _build_cite(cite)
+    documentation = Path(documentation)
+    covers = [spec.shared.fields.URI().deserialize(uri) for uri in covers]
+    test_inputs = [spec.shared.fields.URI().deserialize(uri) for uri in test_inputs]
+    test_outputs = [spec.shared.fields.URI().deserialize(uri) for uri in test_outputs]
+
+    model = model_spec.raw_nodes.Model(
+        format_version=format_version,
+        name=name,
+        description=description,
+        authors=authors,
+        cite=cite,
+        tags=tags,
+        license=license,
+        documentation=documentation,
+        covers=covers,
+        timestamp=timestamp,
+        weights=weights,
+        inputs=[inputs],
+        outputs=[outputs],
+        test_inputs=test_inputs,
+        test_outputs=test_outputs,
+        **kwargs,
+    )
+
+    # serialize and deserialize the raw_nodes.Model to
+    # check that all fields are well formed
+    serialized = model_spec.schema.Model().dump(model)
+    model = model_spec.schema.Model().load(serialized)
+
+    return model
+
+
+def add_weights(model, weight_uri: str, root: Optional[str] = None, weight_type: Optional[str] = None, **weight_kwargs):
+    """Add weight entry to bioimage.io model."""
+    new_weights = _get_weights(weight_uri, weight_type, None, root, **weight_kwargs)[0]
+    model.weights.update(new_weights)
+
+    serialized = model_spec.schema.Model().dump(model)
+    model = model_spec.schema.Model().load(serialized)
+
+    return model
+
+
+def serialize_spec(model, out_path):  # TODO change name to include model (see build_model_spec)
```

Comment:
we can substitute this as well with `bioimageio.spec.save_raw_node`

Reply:
Ok, I will check if that works.

Reply:
removed